### PR TITLE
[NO MRG] Parallelize host serialization/deserialization

### PR DESCRIPTION
**Pull request by jakirkham**
This PR is more intended as an experiment of using threads and streams to speed up device-to-host and host-to-device transfers. In reality we may want some of these changes in other libraries (like CuPy and RMM). Also other ongoing work (like PTDS) may eliminate the need for some of the changes in this PR; thus, simplifying the final result. In any event, hopefully this gives some idea of how we might leverage these features to speed up serialization where a transfer to host is needed (like spilling).

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/abc.py [hunk 1]
```diff
@@ -17,6 +21,36 @@
...
+    else:
+        return memoryview(buf)
+
+
+def _ensure_maybe_on_device(is_cuda, buf):
+    buf = memoryview(buf)
+    if is_cuda:
+        with _get_tls_stream() as stream:
+            #dbuf = rmm.DeviceBuffer(size=buf.nbytes, stream=stream.ptr)
+            dbuf = rmm.DeviceBuffer(size=buf.nbytes)
```
Comment (jakirkham) [inline python/cudf/cudf/core/abc.py:47]: Ideally we would use the `stream.ptr` for allocation of the `DeviceBuffer` here. However when doing that we got a `cudaErrorInvalidResourceHandle` exception (shown in detail below).

Nothing stuck out to me when looking at [the CuPy `Stream` creation code]( https://github.com/cupy/cupy/blob/7f401f0cb67a36ce87ddb5aeb18d38959c9c5b06/cupy/cuda/stream.pyx#L273-L279 ). That said, maybe this isn't too important if this code was rewritten in Cython in RMM or a PTDS-enabled RMM build were used.

In any event, the key takeaway is that deserialization may not be as fast as we would like due to the use of the default stream here.

<details>

```python
/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/site-packages/cudf/core/abc.py in host_deserialize(cls, header, frames)
    156         """
    157         with ThreadPoolExecutor() as executor:
--> 158             frames = list(
    159                 executor.map(
    160                     _ensure_maybe_on_device, header["is-cuda"], frames

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/_base.py in result_iterator()
    609                     # Careful not to keep a reference to the popped future
    610                     if timeout is None:
--> 611                         yield fs.pop().result()
    612                     else:
    613                         yield fs.pop().result(end_time - time.monotonic())

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/_base.py in result(self, timeout)
    430                 raise CancelledError()
    431             elif self._state == FINISHED:
--> 432                 return self.__get_result()
    433 
    434             self._condition.wait(timeout)

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/_base.py in __get_result(self)
    386     def __get_result(self):
    387         if self._exception:
--> 388             raise self._exception
    389         else:
    390             return self._result

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/concurrent/futures/thread.py in run(self)
     55 
     56         try:
---> 57             result = self.fn(*self.args, **self.kwargs)
     58         except BaseException as exc:
     59             self.future.set_exception(exc)

/datasets/jkirkham/miniconda/envs/rapids15dev/lib/python3.8/site-packages/cudf/core/abc.py in _ensure_maybe_on_device(is_cuda, buf)
     44     if is_cuda:
     45         with _get_tls_stream() as stream:
---> 46             dbuf = rmm.DeviceBuffer(size=buf.nbytes, stream=stream.ptr)
     47             cupy.asarray(dbuf).set(numpy.asarray(buf), stream=stream)
     48             return dbuf

rmm/_lib/device_buffer.pyx in rmm._lib.device_buffer.DeviceBuffer.__cinit__()

RuntimeError: CUDA error at: ../include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp365: cudaErrorInvalidResourceHandle invalid resource handle
```

</details>

**Comment by jakirkham** [comment]
Compared the result of serializing/deserializing a largish `DataFrame` both before and after this change. For host serialization this resulted in a 3x speedup in this case. The host deserialization case didn't show much change (if anything this was a bit slower). Though this may be (at least in part) attributable to the use of the default stream for allocation. Using pinned memory may also improve the deserialization case as well.

<details>
<summary>Before this change:</summary>

```python
In [1]: import cupy
   ...: import cudf
   ...: import rmm

In [2]: rmm.reinitialize(pool_allocator=True,
   ...:                  initial_pool_size=int(30 * 2**30))

In [3]: df = cudf.DataFrame({
   ...:     k: cupy.random.random(1_000_000)
   ...:     for i, k in enumerate(map(chr, range(ord("A"), ord("K"))))
   ...: })

In [4]: h, f = df.host_serialize()

In [5]: %timeit df.host_serialize()
32.1 ms ± 253 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [6]: %timeit cudf.DataFrame.host_deserialize(h, f)
10.2 ms ± 20.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

</details>

<details>
<summary>After this change:</summary>

```python
In [1]: import cupy
   ...: import cudf
   ...: import rmm

In [2]: rmm.reinitialize(pool_allocator=True,
   ...:                  initial_pool_size=int(30 * 2**30))

In [3]: df = cudf.DataFrame({
   ...:     k: cupy.random.random(1_000_000)
   ...:     for i, k in enumerate(map(chr, range(ord("A"), ord("K"))))
   ...: })

In [4]: h, f = df.host_serialize()

In [5]: %timeit df.host_serialize()
11.7 ms ± 87.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: %timeit cudf.DataFrame.host_deserialize(h, f)
12.1 ms ± 67.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

</details>

File: python/cudf/cudf/core/abc.py [hunk 2]
```diff
@@ -17,6 +21,36 @@
...
+            cupy.asarray(dbuf).set(numpy.asarray(buf), stream=stream)
```
Comment (jakirkham) [inline python/cudf/cudf/core/abc.py:48]: Doing a deeper dive on the deserialization case, it appears this line takes the bulk of the time in the deserialization case. The `cupy.asarray` and `numpy.asarray` calls are negligible. However calling `.set(...)` takes `~99%` of the time. Computing the transfer rate associated with `.set(...)` in isolation, this comes out to be `~9 GB/s`, which matches what [we have observed before]( https://github.com/rapidsai/dask-cuda/issues/106#issuecomment-664676616 ).

**Comment by jakirkham** [comment]
TL;DR this gets our device-to-host transfer speed up to an equivalent speed as our host-to-device transfers. Both of which are approaching the theoretical limit. As the former limits how fast we can spill from device-to-host, it's good to see that we are approaching PCIe saturation with this relatively simple technique.

**Comment by harrism** [comment]
@jakirkham since this is an experiment, and we don't allow orphan PRs (no project), I need to figure out what project to put this on, or close it. Suggestions?

**Comment by jakirkham** [comment]
cc @pentschev (since you have been looking into PTDS lately 😉)

**Comment by kkraus14** [comment]
moved to 0.18

**Comment by jakirkham** [comment]
Going to close. We can revisit later if needed
